File: ChatGPT/python/output/output_08a.py
# ChatGPT-generated solution will go here.
def main():
    with open('input.txt') as f:
        grid = [line.strip() for line in f if line.strip()]

    rows, cols = len(grid), len(grid[0])
    antennas = {}

    # Collect antenna positions by frequency
    for r in range(rows):
        for c in range(cols):
            ch = grid[r][c]
            if ch != '.':
                antennas.setdefault(ch, []).append((r, c))

    antinodes = set()

    for freq, positions in antennas.items():
        n = len(positions)
        for i in range(n):
            r1, c1 = positions[i]
            for j in range(n):
                if i == j:
                    continue
                r2, c2 = positions[j]
                # Check if (r2, c2) is twice as far from the midpoint as (r1, c1)
                mid_r = 2 * r2 - r1
                mid_c = 2 * c2 - c1
                if 0 <= mid_r < rows and 0 <= mid_c < cols:
                    antinodes.add((mid_r, mid_c))

    print(len(antinodes))

if __name__ == '__main__':
    main()

File: ChatGPT/clojure/output/output_08a.clj
;; ChatGPT-generated solution will go here.
(ns output_08a
  (:require [clojure.string :as str]
            [clojure.set :as set]))

(defn parse-antennas [lines]
  (let [rows (count lines)
        cols (count (first lines))]
    (for [r (range rows)
          c (range cols)
          :let [ch (get-in lines [r c])]
          :when (not= ch \.)]
      [ch [r c]])))

(defn in-bounds? [[r c] lines]
  (and (<= 0 r) (< r (count lines))
       (<= 0 c) (< c (count (first lines)))))

(defn midpoint [[r1 c1] [r2 c2]]
  [(quot (+ r1 r2) 2) (quot (+ c1 c2) 2)])

(defn delta [[r1 c1] [r2 c2]]
  [(- r2 r1) (- c2 c1)])

(defn add-pos [[r c] [dr dc]]
  [(+ r dr) (+ c dc)])

(defn part [lines]
  (let [grid (mapv vec lines)
        antennas (parse-antennas grid)
        by-freq (group-by first antennas)
        antinode-positions
        (->> by-freq
             (mapcat (fn [[_ freqs]]
                       (let [positions (map second freqs)]
                         (for [[a b] (for [i (range (count positions))
                                           j (range (inc i) (count positions))]
                                       [(nth positions i) (nth positions j)])
                               :let [mid (midpoint a b)
                                     d (delta a b)
                                     a2 (add-pos mid d)
                                     b2 (add-pos mid (map - d))]
                               :when (= (add-pos a2 b2) (map + a b))]
                           (filter #(in-bounds? % grid) [a2 b2])))))
             (apply set/union))]
    (count antinode-positions)))

(defn -main [& args]
  (let [filename (first args)
        lines (str/split-lines (slurp filename))]
    (println (part lines))))
